# Day 1 Exercises
## Data Types

#### 1. What types are involved in the following expressions? What data type will each expression evaluate to?


In [55]:
8 * 2.5

20.0

2 data types are present. '8' represents Integer Number data. '2.5' represents Floating Point Number data. The operator is the '*' commanding a multiplication action on the operands of '8' and '2.5'. The expression will evaluate to '20'

In [56]:
9 / 2

4.5

Only the integer data type is present through the '9' and '2'. The operator is the '/' commanding a division action on the operands of '9' and '2'. The expression will evaluate to '4.5'

In [57]:
1.5 * 2 >= 7 - 3

False

2 data types are present: Integer ('2','7','3') and Float ('2.5'). The operators are the '*', '-', and as a boolean operator '>=' (greater than or equal to). The operands are '1.5', '2', '7', and '3'. The expression will evaluate to 'False' on the basis that '1.5*2' is not greater than or equal to '7-3'.

#### 2. In which order will the expressions be evaluated?

In [58]:
6 * 3 + 7 * 4

46

It will be evaluated like so: multiplication then addition. It will evaluate to 46.

In [59]:
5 - 2 * 3 ** 4

-157

It will be evaluated like so: Exponentiation, multiplication, then subtraction. it will evaluate to -157.

In [60]:
(5 - 2) * 3 ** 4

243

It will be evaluated like so: Parenthesis, exponentiation, multiplication, then subtraction. It will evaluate to 243

In [61]:
5 + 2 >= 3 * 4

False

It will be evaluated like so: Multiplication, addition, then the comparative operator '>='. It will evaluate as False

## Comments and Errors

#### 1. Which of these expressions results in an error? Why does each error occur?

In [62]:
((((5 * 4 ** 7))))

81920

This is a valid expression. 4 to the 7th power, multiplied by 5.

In [63]:
84 *= 0.5 / 7

SyntaxError: 'literal' is an illegal expression for augmented assignment (868797654.py, line 1)

This expression is invalid, making it produce a SyntaxError. The operator '*=' is supposed to be used in the circumstance in which a variable (or string) is assigned a value (lets say x=2), multipled by another value (for example, 3), and then the result is reincorporated into the initial variable. Using the example just gone through, x would then be 6. In the invalid expression above, 84 is not an assigned variable; to which if we said, x=84, then x*= 0.5/7, that would then be a valid expression.

In [ ]:
(-(-(-(-5 * (4 + 3)))))

35

This is a valid expression. The evaluate the inner most parenthesis (7), multiply by -5 (-35), then acccount for the 3 negations which when multiplied with the -35, produce an even integer of 35.

In [ ]:
5 * 3 = weight

SyntaxError: cannot assign to operator (3700621706.py, line 1)

This is an invalid expression causing a SyntaxError since you are assigning an expression to a variable; to correct, assign the variable to the expression (i.e., weight = 5*3) Ergo, when you prescribe 'weight', now it will produce a result of '15'. 

In [ ]:
73 / -----------5

-14.6

This is a valid expression. The several '-'s are simply negation signs, which, in theory, are simply -1 mulitiplying an odd amount of times to produce a '-5' of which can then be used as the divisor to 73.

#### 2. Write a block of code to check if the value is a string. If it is, print out a message saying it is valid. Otherwise, raise an exception with a message saying the value must be a string.

In [ ]:
value = 6
if isinstance(value, str):
    print('value is string')
else: print('value must be a string')




value must be a string


#### 3. What are the following errors?

In [ ]:
arr = [1, 2, 3]
print(arr[3])

IndexError: list index out of range

This expression is producing an IndexError since the called index [3] is non-existant. Python functions to index the first character as 0, ergo, what we percieve as the 3rd (r), is, according to Python, 2nd. Change code to;

print(arr[2])

In [ ]:
fullName = "Taylor Swift"
print(full_name)

Taylor Swift


I had issues with this question (initially an error displayed but then it disappeared). I believe this would have been a syntax error since 'fullName' is not equivalent to 'full_name'; they are (would be) case sensitive. 

In [ ]:
2 + '2' + 4

8

This expression produces a TypeError since it is mistakenly combining data types (integers with string). To fix, remove the quotes around '2', and make the expression: 2 + 2 + 4

## Functions

#### 1. Following the function design recipe, define a function that converts kilometers into miles. (Assume there are 1.6 kilometers in a mile.)

In [ ]:
def km_to_m(km):
    '''
    Converting Kilometres to Miles.

    Parameters:
    km (float): the distance in kilometres.

    Returns:
    m (float): the distance in miles.
    '''
    return (km * 0.6)

#this is a test#
km_to_m(10)
#it works!


6.0

#### 2. What value is printed in the below code?

In [64]:
answer = 3

def answer_to_everything():
    '''Return the answer to life, the universe, and everything'''
    answer = 42
    return answer

answer_to_everything()

print(answer)

3


The value of '3' is printed in the above code since it is globally defined. The answer of '42' is defined and would be produced locally within or whenever the function 'answer_to_everything' is called. But generally speaking, since 3 is globally defined as 'answer' and vice versa, when the code runs, it will produce answer = 3, so 3.

#### 3. Complete the body of the function. You can assume num is always >= 0.

In [ ]:
def repeat(string, num):
    ''' Return string repeated num times.
    >>> repeat('yes', 4)
    'yesyesyesyes'
    >>> repeat('no', 0)
    ' '
    >>> repeat('yesnomaybe', 3)
    'yesnomaybeyesnomaybeyesnomaybe'
    '''

    return string * num

#this is a test#
repeat('yes', 10)
#it works!

'yesyesyesyesyesyesyesyesyesyes'

#### 4. Complete the function below. You can assume that number will always be a string with 10 digits and that the first 3 digits will always be the area code.

In [ ]:
def is_416_number(number):
    ''' Check whether the number has a 416 area code.
    >>> is_416_number('(416)-555-5555')
    True
    >>> is_416_number('514 416 5555')
    False
    >>> is_416_number('4165554160')
    True
    '''
    # Your code goes here
    # Hint: replace () and spaces with nothing

SyntaxError: invalid syntax (1541054713.py, line 11)

In [ ]:
#my code goes here

def is_416_number(number):
    if number[0:3] == '416':
        return True
    else:
        return False
    
        
#testing function
is_416_number('5144165555')
#it works!

False